sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/cardinality/ApproximateDistinct.java

```diff
@@ -0,0 +1,352 @@
+   * optimize memory and improve accuracy for small cardinalities.
+   * By calling this builder, you will not use the sparse representation.
+   * If you want to, see {@link ApproximateDistinctFn#withSparseRepresentation(int)}
+   *
+   * <p>Example of use
+   * <pre> {@code PCollection<KV<Integer, String>> input = ...;
+   * PCollection<KV<Integer, HyperLogLogPlus>> hllSketch = input
+   *        .apply(ApproximateDistinct.<Integer, String>perKey(15));
+   * } </pre>
+   *
+   * @param p       number of bits for indexes in the HyperLogLogPlus.
+   *
+   */
+  public static <K, InputT> Combine.PerKey<K, InputT, HyperLogLogPlus> perKey(
+      int p) {
+    return Combine.<K, InputT, HyperLogLogPlus>perKey(ApproximateDistinctFn.<InputT>create(p));
+  }
+
+    /**
+     * Do the same as {@link ApproximateDistinct#globally(int)}, but with a default value for p.
+     *
+     * @param <K>       the type of the keys in the input and output {@code PCollection}s
+     * @param <InputT>  the type of values in the input {@code PCollection}
+     */
+  public static <K, InputT> Combine.PerKey<K, InputT, HyperLogLogPlus> perKey() {
+    return perKey(18);
+  }
+
+  /**
+   * A {@code Combine.CombineFn} that computes the stream into a {@link HyperLogLogPlus}
+   * sketch, useful as an argument to {@link Combine#globally} or {@link Combine#perKey}.
+   *
+   * <p>The HyperLogLog algorithm relies on the principle that the overall cardinality
+   * can be estimated thanks to the longest run of starting 0s of the hashed elements.
+   * The longer the run is, the more unlikely it was to happen. Thus, the greater the number
+   * of elements that have been hashed before getting such a run.
+   * <br>Because this algorithm relies mainly on randomness, the stream is divided into buckets
+   * in order to reduce the variance of the estimation.
+   * Therefore, an estimation is applied on several samples and the overall estimation
+   * is then computed using an average (Harmonic mean).
+   *
+   * @param <InputT>    the type of the elements being combined
+   */
+  public static class ApproximateDistinctFn<InputT>
+      extends Combine.CombineFn<InputT, HyperLogLogPlus, HyperLogLogPlus> {
+
+    private final int p;
+
+    private final int sp;
+
+    private ApproximateDistinctFn(int p, int sp) {
+      this.p = p;
+      this.sp = sp;
+    }
+
+      /**
+       * Returns an {@code ApproximateDistinctFn} combiner with the given precision value p.
+       * This means that the input elements will be dispatched into 2^p buckets
+       * in order to estimate the cardinality.
+       *
+       * @param p           precision value for the normal representation
+       * @param <InputT>    the type of the Input {@code Pcollection}'s elements being combined.
+       */
+    public static <InputT> ApproximateDistinctFn<InputT> create(int p) {
+      return new ApproximateDistinctFn<>(p, 0);
```

Comment:
<!--new_thread; commit:6bf51fdf6afe6f389415792f335dd08adf9fa85a; resolved:0-->
Perform validation that p >= 4 && p <= 32